### PR TITLE
Do not exit and re-enter main event loop during shutdown

### PR DESCRIPTION
**Pull request by hebasto**
On master (1ef34ee25ed34b2b092f15bf3dca5c0508092829) during shutdown `QApplication` exits the main event loop, then re-enter again.

This PR streamlines shutdown process by removing the need to interrupt the main event loop, that is required for #59.

Also, blocking [`QDialog::exec()`](https://doc.qt.io/qt-5/qdialog.html#exec) calls are replaced with safer [`QDialog::show()`](https://doc.qt.io/qt-5/qwidget.html#show), except for `SendConfirmationDialog` as that change is not trivial (marked as TODO).

The [`QDialog::open()`](https://doc.qt.io/qt-5/qdialog.html#open) was not used because the actual modality mode (application modal or window modal) of a dialog depends on whether it has a parent.

This PR does not change behavior, and all touched dialogs are still application modal.
As a follow up, a design research could suggest to make some dialogs window modal.

NOTE for reviewers: quitting app while a dialog is open (e.g., via systray icon menu) must work fine.

### CODE REVIEW

**Review by jarolrod — COMMENTED** [review]
Concept ACK, Tested that each commit can compile and run on its own.

Testing on macOS, some of the dialogs will now open in a subwindow instead of its own window. Here is an example of this:

| Master        | PR               |
| ----------- | ----------- |
| Screen Shot 2021-05-17 at 7 01 48 PM | Screen Shot 2021-05-17 at 7 00 06 PM |

Specifically the `OptionsDialog`, `EditAddressDialog`, `AskPassphraseDialog` undergo this change compared to master. Interestingly, `CoinControlDialog`, `PSBTOperationsDialog`, and `HelpMessageDialog` all still open up in their own window. It is surprising to me that one set of dialogs behave differently from another set under these changes.

**Comment by hebasto** [comment]
Updated 09e84de6bbe550d7e1cc3d56efbded73de36dc38 -> 77a926101d91c4e4c0cead97272d99dbc18285b5 ([pr336.02](https://github.com/hebasto/gui/commits/pr336.02) -> [pr336.03](https://github.com/hebasto/gui/commits/pr336.03), [diff](https://github.com/hebasto/gui/compare/pr336.02..pr336.03)):

- addressed @jarolrod's [comment](https://github.com/bitcoin-core/gui/pull/336#pullrequestreview-661516031)
- fixed bugs

**Comment by hebasto** [comment]
Rebased 77a926101d91c4e4c0cead97272d99dbc18285b5 -> 0f2fff335a28cc287a1e31f9131d1ce1dc8a4739 ([pr336.03](https://github.com/hebasto/gui/commits/pr336.03) -> [pr336.04](https://github.com/hebasto/gui/commits/pr336.04)) on top of the recent CI changes.

**Comment by hebasto** [comment]
Added a note for reviewers to the OP.

File: src/qt/askpassphrasedialog.cpp [hunk 1]
```diff
@@ -70,6 +70,9 @@ AskPassphraseDialog::AskPassphraseDialog(Mode _mode, QWidget *parent, SecureStri
     connect(ui->passEdit3, &QLineEdit::textChanged, this, &AskPassphraseDialog::textChanged);
 
     GUIUtil::handleCloseWindowShortcut(this);
+
+    setAttribute(Qt::WA_DeleteOnClose);
```
Comment (promag) [inline src/qt/askpassphrasedialog.cpp:74]: For the record, I've suggested to add `GUIUtil::showModalDialogAndDeleteOnClose` instead.

**Comment by hebasto** [comment]
Updated 0f2fff335a28cc287a1e31f9131d1ce1dc8a4739 -> 2ad7b4d100fbf41d4d7904e6fbeeec39597c9201 ([pr336.04](https://github.com/hebasto/gui/commits/pr336.04) -> [pr336.05](https://github.com/hebasto/gui/commits/pr336.05)):

- addressed @promag's [comment](https://github.com/bitcoin-core/gui/pull/336#discussion_r646698199)
- rebased on top of the recent GUI changes (signal-slot connection etc)

File: src/qt/askpassphrasedialog.cpp (hunk 1 above)
Comment (hebasto) [reply src/qt/askpassphrasedialog.cpp]: Thanks! [Updated](https://github.com/bitcoin-core/gui/pull/336#issuecomment-856061619).

File: src/qt/guiutil.h [hunk 2]
```diff
@@ -417,6 +418,13 @@ namespace GUIUtil
             type);
     }
 
+    /**
+     * Shows a QDialog instance asynchronously, and deletes it on close.
+     */
+    void ShowModalDialogAndDeleteOnClose(
+        QDialog* dialog,
+        Qt::WindowModality modality = Qt::ApplicationModal);
```
Comment (promag) [inline src/qt/guiutil.h:426]: Why this as an argument when it's not used? If you keep this then the function could be `ShowDialogAndDeleteOnClose`?

Comment (hebasto) [reply src/qt/guiutil.h:426]: It could be `Qt::WindowModal` as well.

Comment (hebasto) [reply src/qt/guiutil.h:426]: > Why this as an argument when it's not used?

For generality.

**Comment by hebasto** [comment]
Updated 2ad7b4d100fbf41d4d7904e6fbeeec39597c9201 -> b441b47b046e9952a6bbc582dc5fa6044961573b ([pr336.05](https://github.com/hebasto/gui/commits/pr336.05) -> [pr336.06](https://github.com/hebasto/gui/commits/pr336.06)):

- rebased due to the conflict with #4
- addressed @promag's [comment](https://github.com/bitcoin-core/gui/pull/336#discussion_r646732102)

File: src/qt/guiutil.h (hunk 2 above)
Comment (hebasto) [reply src/qt/guiutil.h]: Thanks! [Updated](https://github.com/bitcoin-core/gui/pull/336#issuecomment-859333144).

**Comment by hebasto** [comment]
Rebased b441b47b046e9952a6bbc582dc5fa6044961573b -> d6dd44c69428b641609c2564ed39b97fdef7be92 ([pr336.06](https://github.com/hebasto/gui/commits/pr336.06) -> [pr336.07](https://github.com/hebasto/gui/commits/pr336.07)) due to the conflict with #381.

**Comment by hebasto** [comment]
Rebased d6dd44c69428b641609c2564ed39b97fdef7be92 -> 950ed3cb529f78255614f2ffa148b04c77e61c2d ([pr336.07](https://github.com/hebasto/gui/commits/pr336.07) -> [pr336.08](https://github.com/hebasto/gui/commits/pr336.08)) due to the conflict with #399.

**Comment by hebasto** [comment]
Rebased 2407f0d7e0771d6edbaba0b25510cf0448ad2563 -> 94c395dd815f69c244044a0e772ea443a1b46c8c ([pr336.09](https://github.com/hebasto/gui/commits/pr336.09) -> [pr336.10](https://github.com/hebasto/gui/commits/pr336.10)) due to the conflict with #398.

**Comment by laanwj** [comment]
Concept ACK

File: src/qt/bitcoingui.h [hunk 3]
```diff
@@ -214,6 +214,7 @@ class BitcoinGUI : public QMainWindow
     void openOptionsDialogWithTab(OptionsDialog::Tab tab);
 
 Q_SIGNALS:
+    void quitClicked();
```
Comment (promag) [inline src/qt/bitcoingui.h:217]: 1d4760cc6b939f22f60b24c4989f55c7c4fa79d8

nit, quitTriggered or quitRequested.

**Review by promag — CHANGES_REQUESTED** [review]
Tested ACK 94c395dd815f69c244044a0e772ea443a1b46c8c.

I've tested quitting in different cases, with some dialogs opened, also with the stop RPC command.

Checked all call sites affected by the change to non-blocking.

**Comment by hebasto** [comment]
Updated 94c395dd815f69c244044a0e772ea443a1b46c8c -> 451ca244db8bc71ffc3cc9982d025f144cc8f3bc ([pr336.10](https://github.com/hebasto/gui/commits/pr336.10) -> [pr336.11](https://github.com/hebasto/gui/commits/pr336.11), [diff](https://github.com/hebasto/gui/compare/pr336.10..pr336.11)):

- addressed @promag's [comment](https://github.com/bitcoin-core/gui/pull/336#discussion_r718524353)

File: src/qt/bitcoingui.h (hunk 3 above)
Comment (hebasto) [reply src/qt/bitcoingui.h]: Thanks! [Updated](https://github.com/bitcoin-core/gui/pull/336#issuecomment-930216684).

**Review by promag — COMMENTED** [review]
ACK 451ca244db8bc71ffc3cc9982d025f144cc8f3bc, just changed signal to `quitRequested`.

**Comment by hebasto** [comment]
@shaavan 

> tACK [2407f0d](https://github.com/bitcoin-core/gui/commit/2407f0d7e0771d6edbaba0b25510cf0448ad2563)
> Tested on Ubuntu 20.04 (Qt version 5.12.8)
> I tested all the commits and didn’t observe any changes in functionality.

Would you mind re-reviewing this PR after the recent changes?

**Comment by shaavan** [comment]
> Would you mind re-reviewing this PR after the recent changes?

Let me take a look!

**Comment by laanwj** [comment]
Code review and lighly tested ACK 451ca244db8bc71ffc3cc9982d025f144cc8f3bc

**Review by shaavan — COMMENTED** [review]
Post-merge ACK 451ca244db8bc71ffc3cc9982d025f144cc8f3bc

Changes since my last review:
- PR was rebased on Master
- Function `QuitClicked` is renamed to `QuitRequested`

Tested successfully again on Ubuntu 20.04 (Using Qt version 5.12.8), with no change in functionality. Thanks for this fantastic work @hebasto!

**Comment by prayank23** [comment]
@hebasto Can we revert the changes made in this pull request?

**Comment by hebasto** [comment]
@prayank23 
> @hebasto Can we revert the changes made in this pull request?

I'm working on https://github.com/bitcoin/bitcoin/issues/23790

**Comment by hebasto** [comment]
@prayank23 
> Can we revert the changes made in this pull request?

Blocking synchronous dialogs are evil for responsible GUI. Only broken part was reverted in #509.

**Comment by prayank23** [comment]
> Blocking synchronous dialogs are evil for responsible GUI. Only broken part was reverted in #509.

Compiling PR branch. Will test and comment in PR. Thanks.

**Pull request merged**